src/pip/_internal/utils/direct_url_helpers.py

```diff
@@ -85,3 +85,46 @@ def direct_url_from_link(
             info=ArchiveInfo(hash=hash),
             subdirectory=link.subdirectory_fragment,
         )
+
+
+def _link_from_direct_url(direct_url: DirectUrl) -> Link:
+    """Create a link from given direct URL construct.
+
+    This function is designed specifically for ``link_matches_direct_url``, and
+    does NOT losslessly reconstruct the original link that produced the
+    DirectUrl. Namely:
+
+    * The auth part is ignored (since it does not affect link equivalency).
+    * Only "subdirectory" and hash fragment parts are considered, and the
+      ordering of the kept parts are not considered (since only their values
+      affect link equivalency).
+
+    .. seealso:: ``pip._internal.models.link.links_equivalent()``
+    """
+    url = direct_url.url
+    hash_frag: Optional[str] = None
+
+    direct_url_info = direct_url.info
+    if isinstance(direct_url_info, VcsInfo):
+        url = f"{direct_url_info.vcs}+{url}"
+        if direct_url_info.requested_revision is not None:
+            url = f"{url}@{direct_url_info.requested_revision}"
```

Comment:
```suggestion
        url = f"{direct_url_info.vcs}+{url}@{direct_url_info.commit_id}"
```

I think we should compare the commit id for with what is already installed, and the requested revision of what is being requested for installation. This way if we install, say, a vcs tag, we record the commit id, and then when we install with the commit id, it considers it is already installed. But if we install with the tag again, it reinstalls.

Reply:
What is the difference between `commit_id` and `resolved_revision` here?